### PR TITLE
fix: use temp dir for secret data

### PR DESCRIPTION
**Pull request by dimaqq**
Change the mechanism used to pass secret values from ops to hook tools from command line to files in temporary directory.

Removes the 2MB total secret size limit imposed by command line length

Hook tool calls juju agent which in turn reads these files, I've confirmed with the Juju team that juju agent and charm code are run as the same user.

### CODE REVIEW

**Comment by dimaqq** [comment]
Also tested manually to replacing ops in a running charm venv.

**Review by benhoyt — APPROVED** [review]
Yes, I like this approach a lot better, thanks! The code reads a lot more simply, and it's nice to avoid keeping N files open. If you've done manual testing, I'm good with this.

File: ops/model.py [hunk 1]
```diff
@@ -3598,11 +3598,13 @@ def secret_set(
...
             args += ['--rotate', rotate.value]
-        if content is not None:
-            # The content has already been validated with Secret._validate_content
-            for k, v in content.items():
-                args.append(f'{k}={v}')
-        self._run_for_secret('secret-set', *args)
+        with tempfile.TemporaryDirectory() as tmp:
+            # The content is None or has already been validated with Secret._validate_content
+            for k, v in (content or {}).items():
+                with open(f'{tmp}/{k}', mode='w', encoding='utf-8') as f:
```
Comment (benhoyt) [inline ops/model.py:3604]: At first I was concerned about slashes in keys, but we (and Juju) is very strict about keys (basically numbers and letters and hyphens) and they've already been checked.

**Review by tonyandrewmeyer — APPROVED** [review]
Did a quick check (Juju 3.6, LXD) and all seems good here too. Agreed that this approach seems generally nicer.

File: ops/model.py (hunk 1 above)
Comment (dimaqq) [reply ops/model.py:3604]: Indeed.
I've checked that actually.
The only possible unaccounted for is a key that's very very long, exceeding directory entry name limit, as we don't limit the length in ops.
I figure it's OK to fail in that case.

**Comment by dimaqq** [comment]
Fixes CVE-2024-41129

**Pull request merged**